core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java

```diff
@@ -0,0 +1,504 @@
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * Class for managing peristent inode tree state.
+ *
+ * This class owns all persistent inode tree state, and all inode tree modifications must go through
+ * this class. The getInodesView and getRoot methods expose unmodifiable views of the inode tree.
+ * To modify the inode tree, create a journal entry and call one of the applyAndJournal methods.
+ */
+public class InodeTreePersistentState {
+  private static final Logger LOG = LoggerFactory.getLogger(InodeTreePersistentState.class);
+
+  private static final IndexDefinition<Inode<?>, Long> ID_INDEX =
+      new IndexDefinition<Inode<?>, Long>(true) {
+        @Override
+        public Long getFieldValue(Inode<?> o) {
+          return o.getId();
+        }
+      };
+
+  /** All inodes, indexed by ID. */
+  private final FieldIndex<Inode<?>, Long> mInodes = new UniqueFieldIndex<>(ID_INDEX);
+  /** The root of the entire file system. */
+  private InodeDirectory mRoot = null;
+
+  /**
+   * A set of inode ids representing pinned inode files. These are not part of the journaled state,
+   * but this class keeps the set of pinned inodes up to date whenever the inode tree is modified.
+   *
+   * This class owns this set, and no other class can modify the set.
+   */
+  private final Set<Long> mPinnedInodeFileIds = new ConcurrentHashSet<>(64, 0.90f, 64);
+
+  /**
+   * TTL bucket list. The list is owned by InodeTree, and is only shared with
+   * InodeTreePersistentState so that the list can be updated whenever inode tree state changes.
+   */
+  private final TtlBucketList mTtlBuckets;
+
+  /** Unmodifiable view of mInodes. */
+  private final InodesView mInodesView = new InodesView(mInodes);
+
+  /**
+   * @param ttlBucketList reference to the ttl bucket list so that the list can be updated when the
+   *        inode tree is modified
+   */
+  public InodeTreePersistentState(TtlBucketList ttlBucketList) {
+    mTtlBuckets = ttlBucketList;
+  }
+
+  /**
+   * @return the inodes of the inode tree, indexed by id
+   */
+  public InodesView getInodesView() {
+    return mInodesView;
+  }
+
+  /**
+   * @return the root of the inode tree
+   */
+  public InodeDirectoryView getRoot() {
+    return mRoot;
+  }
+
+  /**
+   * @return the pinned inode file ids;
+   */
+  public Set<Long> getPinnedInodeFileIds() {
+    return Collections.unmodifiableSet(mPinnedInodeFileIds);
+  }
+
+  /**
+   * Applies a journal entry to the inode tree state. This method should only be used during journal
+   * replay. Otherwise, use one of the applyAndJournal methods.
+   *
+   * @param entry the entry
+   */
+  public void replayJournalEntryFromJournal(JournalEntry entry) {
+    if (entry.hasDeleteFile()) {
```

Comment:
I've consolidated it so that there is only one place that does the checks. For covering this, I think we should add a test which creates at least one of each type of journal entry, then replays the journal. I'll add this in a followup PR.